tiledb/sm/c_api/tiledb.cc

```diff
@@ -3640,13 +3640,17 @@ int32_t tiledb_subarray_alloc(
         ctx->ctx_->storage_manager());
     (*subarray)->is_allocated_ = true;
   } catch (...) {
+    auto st = Status_Error("Failed to construct TileDB subarray object");
+    LOG_STATUS(st);
+    save_error(ctx, st);
+    return TILEDB_ERR;
   }
   if ((*subarray)->subarray_ == nullptr) {
     delete *subarray;
     auto st = Status_Error("Failed to allocate TileDB subarray object");
     LOG_STATUS(st);
     save_error(ctx, st);
-    return TILEDB_OOM;
+    return TILEDB_ERR;
```

Comment:
Isn't this technically still a failed allocation?
- we use `std::nothrow` for the constructor
- but other operations in the block above could throw, and will be caught by `catch (...)`
- we should be left with only `bad_alloc` as the only reason for `== nullptr`

I think the following would improve the code structure and error reporting upon failure:

```
try {
...
} catch (bad_alloc) {
// existing "Failed to allocate" message here
return TILEDB_OOM;
} catch (exception e) {
// log existing `"Failed to allocate TileDB subarray object"` PLUS e.what()
return TILEDB_ERR;
}
```

This way we give precise error messages for most exceptions, and we remove the fallthrough `catch (...)` which currently may the wrong reason for failure (if the exceptions is _not_ due to OOM). We can leave the case of `catch (...)` to the api_entry wrapper.

Reply:
If this PR merges after #3222, we don't need `nothrow` any more; the API wrapper will do the right thing. Other error handling is simpler as well, since we can throw `CAPIStatusException` in one line.